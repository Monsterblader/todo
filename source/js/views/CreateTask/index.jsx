import React, { Component } from "react";

export default class CreateTask extends Component {
  state = {
    task: "",
    startDate: "",
    endDate: "",
  };

  handleChange = e => {
    const target = e.target;
    const name = target.name;

    this.setState({[name]: target.value});
  }
  ;

      handleKeyPress = e => {
        if (e.key !== "Enter")
          return;

        const {onSubmitEditing} = this.props;
        const {task, startDate, endDate} = this.state;

        if (!task)
          return; // Don't submit if empty

        onSubmitEditing({task, startDate, endDate});
        this.setState({
          endDate: "",
          index: "",
          parent: "",
          startDate: "",
          task: "",
        });
      }
  ;

      render() {
    const {placeholder} = this.props;
    const {task, startDate, endDate} = this.state;

    return (
        <div>
          <input
            name="task"
            style={styles.input}
          type={"text"}
          value={task}
          placeholder={placeholder}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
          />
          <input
            name="startDate"
            style={styles.input}
          type={"date"}
          value={startDate}
          placeholder={placeholder}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
          />
        <input
          name="endDate"
          style={styles.input}
          type={"date"}
          value={endDate}
          placeholder={placeholder}
          onChange={this.handleChange}
          onKeyPress={this.handleKeyPress}
          />
        </div>
        );
  }
}

const styles = {
  input: {
    fontSize: "100%",
    padding: 15,
    borderWidth: 0
  }
};

