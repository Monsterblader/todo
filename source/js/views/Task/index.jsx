import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {findDOMNode} from 'react-dom';
import {DragSource, DropTarget} from 'react-dnd';
import {ItemTypes} from 'Constants';

const taskSource = {
  beginDrag(props) {
    return {id: props.id, index: props.index, value: props.value};
  },

  endDrag(props, monitor, component) {
    if (!monitor.didDrop()) {
      return;
    }

    const item = monitor.getItem();
    const dropResult = monitor.getDropResult();

    // This is a good place to call some Flux action
    //    CardActions.moveCardToList(item.id, dropResult.listId);
  }
};

const taskTarget = {
  hover(props, monitor, component) {
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    if (dragIndex === hoverIndex) {
      return;
    }

    const hoverBoundingRect = findDOMNode(component).getBoundingClientRect();
    const hoverMiddleY = (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
    const clientOffset = monitor.getClientOffset();
    const hoverClientY = clientOffset.y - hoverBoundingRect.top;

    if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
      return;
    }

    if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
      return;
    }

    props.moveTask(dragIndex, hoverIndex);
    monitor.getItem().index = hoverIndex;
  }
};

function dragCollect(connect, monitor) {
  return {connectDragSource: connect.dragSource(), isDragging: monitor.isDragging()};
}

function dropCollect(connect, monitor) {
  return {connectDropTarget: connect.dropTarget(), isOver: monitor.isOver(), didDrop: monitor.didDrop(), getDropResult: monitor.getDropResult()};
}

// @DragSource(ItemTypes.TASK, taskSource, dragCollect)
// @DropTarget(ItemTypes.TASK, taskTarget, dropCollect)
export class Task extends Component {
  static propTypes = {
    value: PropTypes.object.isRequired,
    connectDragSource: PropTypes.func.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
    isDragging: PropTypes.bool
  };

  render() {
    const {
      connectDragSource,
      connectDropTarget,
      isDragging,
      indentTask,
      outdentTask
    } = this.props;
    const {
      index,
      value
    } = this.props;

    const handleIndent = () => {
      const source = this.props.index;

      indentTask(source);
    };

    const handleOutdent = () => {
      const source = this.props.index;

      outdentTask(source);
    }

    /* render sub-tasks */
    return connectDragSource(connectDropTarget(
      <div id={index} style={{
        opacity: isDragging
          ? 0.5
          : 1,
        fontSize: 25,
        fontWeight: 'bold',
        cursor: 'move'
      }} className='singleTask'>
        {index !== 0 && <div className='outdentTask' onClick={handleOutdent}>Outdent</div>}
        {index !== 0 && <div className='indentTask' onClick={handleIndent}>Indent</div>}
        <div>
          <span>{value.get('task')}</span>
          <span>{value.get('startDate')}</span>
          <span>{value.get('endDate')}</span>
        </div>

        <hr/>
      </div>
    ));
  }
}

export default DragSource(ItemTypes.TASK, taskSource, dragCollect)(DropTarget(ItemTypes.TASK, taskTarget, dropCollect)(Task));
