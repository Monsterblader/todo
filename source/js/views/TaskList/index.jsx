import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Task from 'views/Task';

export class TaskList extends Component {
  static propTypes = {
    taskList: PropTypes.object
  }

  render() {
    const {
      taskList,
      moveTask,
      indentTask,
      outdentTask
    } = this.props;

    return (
      <div className='taskList'>
        <p>
          Task list
        </p>

        <hr/>

        <div className='Example'>
          {taskList.filter(value => value.get('parent') === "").map((value, key) => (
            <Task
              key={`task${key}`}
              id={value.id}
              index={value.get('index')}
              value={value}
              moveTask={moveTask}
              indentTask={indentTask}
              outdentTask={outdentTask}
            />
          ))}
        </div>

      </div>
    );
  }
};

export default connect(state => { taskList: state.app.get('taskList') })(TaskList);
