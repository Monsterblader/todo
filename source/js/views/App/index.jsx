import React, {Component} from 'react';
import PropTypes from 'prop-types';
import HTML5Backend from 'react-dnd-html5-backend';
import {DragDropContext} from 'react-dnd';

import {addTask, moveTask, indentTask} from 'actions/app';

import CreateTask from 'views/CreateTask';
import TaskList from 'views/TaskList';

@DragDropContext(HTML5Backend)
export default class App extends Component {
  static propTypes = {
    children: PropTypes.object
  };

  onAddTask = task => {
    const {dispatch} = this.props.store;

    dispatch(addTask(task));
  };

  onMoveTask = (dragTask, hoverTask) => {
    const {dispatch} = this.props.store;

    dispatch(moveTask(dragTask, hoverTask));
  };

  onIndentTask = (indent, target) => {
    const {dispatch} = this.props.store;

    dispatch(indentTask(indent, target));
  };

  render() {
    return (
      <div className='App'>
        <CreateTask onSubmitEditing={this.onAddTask} placeholder="Task description"/>
        <TaskList moveTask={this.onMoveTask} indentTask={this.onIndentTask}/>
      </div>
    );
  }
}
