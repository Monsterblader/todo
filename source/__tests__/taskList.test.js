import React from 'react';
import { shallow, mount, render } from 'enzyme';
import { TaskList } from '../js/views/TaskList';
import { fromJS } from 'immutable';
import renderer from 'react-test-renderer';

describe('TaskList', () => {
  const mockTaskList = fromJS([
    {
      id: 'a1',
      index: '0',
      task: 'task1',
      startDate: 'startDate1',
      endDate: 'endDate1'
    }, {
      id: 'b2',
      index: '1',
      task: 'task2',
      startDate: 'startDate2',
      endDate: 'endDate2'
    }, {
      id: 'c3',
      index: '2',
      task: 'task3',
      startDate: 'startDate3',
      endDate: 'endDate3'
    }
  ]);

  it('should render', () => {
    const indentTaskSpy = jest.fn(),
      outdentTaskSpy = jest.fn(),
      createTask = renderer.create(
        <TaskList
          taskList={mockTaskList}
          indentTask={indentTaskSpy}
          outdentTask={outdentTaskSpy}
        />
      );

    expect(createTask).toMatchSnapshot();
  });

  it.skip('should render with "task," "startDate," and "endDate"', () => {
    const createTask = shallow(
        <CreateTask
          onSubmitEditing=''
          placeholder="test"
        />
      );

    expect(createTask.find('input')).toHaveLength(3);
    expect(createTask.find('.enterTask')).toHaveLength(1);
    expect(createTask.find('.enterStartDate')).toHaveLength(1);
    expect(createTask.find('.enterEndDate')).toHaveLength(1);
  });

  it.skip('should field changes', () => {
    const onSubmitEditingSpy = jest.fn(),
      testEvent = {target: {name: 'task', value: 'a'}},
      createTask = shallow(
        <CreateTask
          onSubmitEditing={onSubmitEditingSpy}
          placeholder="test"
        />
      );

    createTask.find('.enterTask').simulate('change', testEvent);
    expect(createTask.instance().state.task).toEqual('a');
    testEvent.target.value = 'b';
    createTask.instance().handleChange(testEvent);
    expect(createTask.instance().state.task).toEqual('b');
  });

  it.skip('should handle key presses', () => {
    const onSubmitEditingSpy = jest.fn(),
      testEvent = {target: {name: 'task', value: 'a'}},
      createTask = shallow(
        <CreateTask
          onSubmitEditing={onSubmitEditingSpy}
          placeholder="test"
        />
      );

    createTask.find('.enterTask').simulate('keypress', {key: 'alt'});
    expect(onSubmitEditingSpy).not.toBeCalled();
    createTask.find('.enterTask').simulate('keypress', {key: 'Enter'});
    expect(onSubmitEditingSpy).not.toBeCalled();

    createTask.find('.enterTask').simulate('change', testEvent);
    createTask.find('.enterTask').simulate('keypress', {key: 'Enter'});
    expect(onSubmitEditingSpy).lastCalledWith(fromJS({task: 'a', startDate: '', endDate: '', parent: '', index: ''}));
  });

  it.skip('should handle key presses', () => {
    const onSubmitEditingSpy = jest.fn(),
      testEvent = {target: { name: 'task', value: 'a' }},
      createTask = shallow(
        <CreateTask
          onSubmitEditing={onSubmitEditingSpy}
          placeholder="test"
        />
      );

    createTask.instance().handleKeyPress({ key: 'Alt' });
    expect(onSubmitEditingSpy).not.toBeCalled();

    createTask.instance().handleKeyPress({ key: 'Enter' });
    expect(onSubmitEditingSpy).not.toBeCalled();

    createTask.find('.enterTask').simulate('change', testEvent);
    createTask.find('.enterTask').simulate('keypress', { key: 'Enter' });
    expect(onSubmitEditingSpy).lastCalledWith(fromJS({ task: 'a', startDate: '', endDate: '', parent: '', index: '' }));
    expect(createTask.instance().state).toEqual({ endDate: '', index: '', parent: '', startDate: '', task: '' });
  });
});
