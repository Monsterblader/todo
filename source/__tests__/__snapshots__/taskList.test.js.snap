// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should render 1`] = `
<div
  className="taskList"
>
  <p>
    Task list
  </p>
  <hr />
  <div
    className="Example"
  />
</div>
`;
